webserver/views/api_compat.py

```diff
@@ -0,0 +1,352 @@
+
+    session = Session.load(sk, api_key)
+    if not session:
+        if not Token.is_valid_api_key(api_key):
+            raise InvalidAPIUsage(10, output_format=output_format)   # Invalid API_KEY
+        raise InvalidAPIUsage(9, output_format=output_format)        # Invalid Session KEY
+
+    lookup = defaultdict(dict)
+    for key, value in data.items():
+        if key == "sk" or key == "token" or key == "api_key" or key == "method":
+            continue
+        matches = re.match('(.*)\[(\d+)\]', key)
+        if matches:
+            key = matches.group(1)
+            number = matches.group(2)
+        else:
+            number = 0
+        lookup[number][key] = value
+
+    if request.form['method'].lower() == "track.updatenowplaying":
+        for i, listen in lookup.iteritems():
+            if 'timestamp' not in listen:
+                listen['timestamp'] = calendar.timegm(datetime.now().utctimetuple())
+
+    # Convert to native payload then submit 'em.
+    listen_type, native_payload = _to_native_api(lookup, data['method'])
+    augmented_listens = insert_payload(native_payload, session.user.name, listen_type=listen_type)
+
+    # With corrections than the original submitted listen.
+    doc, tag, text = Doc().tagtext()
+    with tag('lfm', status="ok"):
+        with tag("nowplaying" if listen_type == "playing_now" else "scrobbles"):
+
+            for origL, augL in zip(lookup.values(), augmented_listens):
+                corr = defaultdict(lambda: "0")
+
+                track = augL['track_metadata']['track_name']
+                if origL['track'] != augL['track_metadata']['track_name']:
+                    corr['track'] = '1'
+
+                artist = augL['track_metadata']['artist_name']
+                if origL['artist'] != augL['track_metadata']['artist_name']:
+                    corr['artist'] = '1'
+
+                ts = augL['listened_at']
+
+                albumArtist = artist
+                if origL.get('albumArtist', origL['artist']) != artist:
+                    corr['albumArtist'] = '1'
+
+                # TODO: Add the album part
+                album = ""
+
+                with tag('scrobble'):
+                    with tag('track', corrected=corr['track']):
+                        text(track)
+                    with tag('artist', corrected=corr['artist']):
+                        text(artist)
+                    with tag('album', corrected=corr['album']):
+                        text(album)
+                    with tag('albumArtist', corrected=corr['albumArtist']):
+                        text(albumArtist)
+                    with tag('timestamp'):
+                        text(ts)
+                    with tag('ignoredMessage', code="0"):
+                        text('')
+
+    return format_response('<?xml version="1.0" encoding="utf-8"?>\n' + yattag.indent(doc.getvalue()),
+                           output_format)
+
+
+def format_response(data, format="xml"):
```

Comment:
This seems like a roundabout way of producing the correct output format. From what I understand, you generate an XML string in each view, and then in this method if the format is json you re-parse the xml and convert it to a dictionary.
I wonder if it would be more understandable to create a python dict in each view instead, and then in this `format_response` method you either convert this dict to XML, or serialise it directly to JSON.